src/psyclone/tests/psyir/transformations/intrinsics/matmul2code_trans_test.py

```diff
@@ -925,3 +955,42 @@ def test_apply_matmat_name_clashes(tmpdir, fortran_reader, fortran_writer):
         "    enddo\n"
         "  enddo\n" in out)
     assert Compile(tmpdir).string_compiles(out)
+
+
+def test_apply_matmat_reordered(tmpdir, fortran_reader, fortran_writer):
+    '''
+    Check the apply method works when the second argument to matmul is a
+    matrix but additional indices are present in any order.
+
+    '''
+
+    #  Extra indices in the inputs
+    psyir = fortran_reader.psyir_from_source(
+        "subroutine my_sub()\n"
+        "  real, dimension(2,4,6) :: jac\n"
+        "  real, dimension(4,6,3) :: jac_inv\n"
+        "  real, dimension(2,3) :: result\n"
+        "  result(:,:) = matmul(jac(:,1,:), jac_inv(2,:,:))\n"
+        "end subroutine my_sub\n")
+    trans = Matmul2CodeTrans()
+    assign = psyir.walk(Assignment)[0]
+    trans.apply(assign.rhs)
+    out = fortran_writer(psyir)
+    print(out)
+    assert (
+        "  real, dimension(2,4,6) :: jac\n"
+        "  real, dimension(4,6,3) :: jac_inv\n"
+        "  real, dimension(2,3) :: result\n"
+        "  integer :: i\n"
+        "  integer :: j\n"
+        "  integer :: ii\n"
+        "\n"
+        "  do j = 1, 3, 1\n"
+        "    do i = 1, 2, 1\n"
+        "      result(i,j) = 0.0\n"
+        "      do ii = 1, 6, 1\n"
+        "        result(i,j) = result(i,j) + jac(i,1,ii) * jac_inv(2,ii,j)\n"
```

Comment:
This is great, well done. Please could you add a similar test for a matrix-vector (e.g., when you have `jac_inv(2,:,arg+3)` as the second argument to matmul. It would also be valuable, as I indicate there (the `arg+3`), to ensure that the indices that aren't full ranges can be anything we like.

Reply:
Addressed in 10d5abbfb758dbaa45439829f974a99504e27227. This was a really good shout because there actually wasn't a coverage test for vectors without the first index being a full range.

To get this to work I changed the condition in the `apply` method to look at the number of full ranges, to determine whether it's a matrix-matrix or matrix-vector application. This also required changes in the matrix-vector application to properly pack the vector to create the array reference.

It's the same code as the old `_create_matrix_ref` so I simply renamed this to `_create_array_ref`.